src/Plutarch/Extra/List.hs

```diff
@@ -209,3 +239,154 @@ pfromList ::
   [Term s a] ->
   Term s (list a)
 pfromList = foldr (\x xs -> pcons # x # xs) pnil
+
+{- | Check if two lists are equal, given a custom equalator. The two list types
+     can be different.
+
+     @since 3.14.1
+-}
+plistEqualsBy ::
+  forall
+    (list1 :: (S -> Type) -> (S -> Type))
+    (list2 :: (S -> Type) -> (S -> Type))
+    (a :: S -> Type)
+    (b :: S -> Type)
+    (s :: S).
+  (PIsListLike list1 a, PIsListLike list2 b) =>
+  Term s ((a :--> b :--> PBool) :--> list1 a :--> list2 b :--> PBool)
+plistEqualsBy = phoistAcyclic $
+  plam $ \eq -> pfix #$ plam $ \self l1 l2 ->
+    pelimList
+      ( \x xs ->
+          pelimList
+            ( \y ys ->
+                -- Avoid comparison if two lists have different length.
+                self # xs # ys #&& eq # x # y
+            )
+            -- l2 is empty, but l1 is not.
+            (pconstant False)
+            l2
+      )
+      -- l1 is empty, so l2 should be empty as well.
+      (pnull # l2)
+      l1
+
+{- | / O(n) /. Remove the first occurance of a value from the given list. Return
+     nothing is said value is not in the list.
+
+     @since 3.14.1
+-}
+pdeleteFirstBy ::
+  forall
+    (a :: S -> Type)
+    (list :: (S -> Type) -> (S -> Type))
+    (s :: S).
+  (PIsListLike list a) =>
+  Term s ((a :--> a :--> PBool) :--> a :--> list a :--> PMaybe (list a))
+pdeleteFirstBy = phoistAcyclic $
+  plam $ \f' x -> plet (f' # x) $ \f ->
+    precList
+      ( \self h t ->
+          pif
+            (f # h)
+            (pjust # t)
+            (pfmap # (pcons # h) # (self # t))
+      )
+      (const pnothing)
+
+{- | Partial version of 'pdeleteBy'.
+
+     @since 3.14.1
+-}
+pdeleteFirstBy' ::
+  forall
+    (a :: S -> Type)
+    (list :: (S -> Type) -> (S -> Type))
+    (s :: S).
+  (PIsListLike list a) =>
+  Term s ((a :--> a :--> PBool) :--> a :--> list a :--> list a)
+pdeleteFirstBy' = phoistAcyclic $
+  plam $ \f' x -> plet (f' # x) $ \f ->
+    precList
+      ( \self h t ->
+          pif
+            (f # h)
+            t
+            (pcons # h #$ self # t)
+      )
+      (const $ ptraceError "Cannot delete element")
```

Comment:
The convention for naming variants that crash instead of returning `PMaybe` or similar is to use `try`. So in this case, it'd be `ptryDeleteFirstBy`.